joint_velocity_example_controller:
    type: franka_example_controllers/JointVelocityExampleController
    joint_names:
        - panda_joint1
        - panda_joint2
        - panda_joint3
        - panda_joint4
        - panda_joint5
        - panda_joint6
        - panda_joint7

joint_position_example_controller:
    type: franka_example_controllers/JointPositionExampleController
    joint_names:
        - panda_joint1
        - panda_joint2
        - panda_joint3
        - panda_joint4
        - panda_joint5
        - panda_joint6
        - panda_joint7

cartesian_pose_example_controller:
    type: franka_example_controllers/CartesianPoseExampleController
    arm_id: panda

elbow_example_controller:
    type: franka_example_controllers/ElbowExampleController
    arm_id: panda

cartesian_velocity_example_controller:
    type: franka_example_controllers/CartesianVelocityExampleController
    arm_id: panda

model_example_controller:
    type: franka_example_controllers/ModelExampleController
    arm_id: panda

force_example_controller:
    type: franka_example_controllers/ForceExampleController
    arm_id: panda
    joint_names:
        - panda_joint1
        - panda_joint2
        - panda_joint3
        - panda_joint4
        - panda_joint5
        - panda_joint6
        - panda_joint7

joint_impedance_example_controller:
    type: franka_example_controllers/JointImpedanceExampleController
    arm_id: panda
    joint_names:
        - panda_joint1
        - panda_joint2
        - panda_joint3
        - panda_joint4
        - panda_joint5
        - panda_joint6
        - panda_joint7
    k_gains:
        - 600.0
        - 600.0
        - 600.0
        - 200.0
        - 250.0
        - 150.0
        - 50.0
    d_gains:
        - 50.0
        - 50.0
        - 50.0
        - 20.0
        - 20.0
        - 20.0
        - 10.0
    radius: 0.1
    acceleration_time: 2.0
    vel_max: 0.15
    publish_rate: 10.0
    coriolis_factor: 1.0

cartesian_impedance_example_controller:
    type: franka_example_controllers/CartesianImpedanceExampleController
    arm_id: panda
    joint_names:
        - panda_joint1
        - panda_joint2
        - panda_joint3
        - panda_joint4
        - panda_joint5
        - panda_joint6
        - panda_joint7
    # # ↓ 在这里添加你的参数 ↓
    translational_x_stiffness:  0.0    # 尽可能低的平移刚度
    translational_y_stiffness:  0.0
    translational_z_stiffness:  0.0

    rotational_x_stiffness:     0.0    # 尽可能低的旋转刚度
    rotational_y_stiffness:     0.0
    rotational_z_stiffness:     0.0

    translational_damping:      0.1    # 非零但极低的平移阻尼
    rotational_damping:         0.1    # 非零但极低的旋转阻尼

    nullspace_stiffness:        0.0    # 关节零空间刚度


dual_arm_cartesian_impedance_example_controller:
    type: franka_example_controllers/DualArmCartesianImpedanceExampleController
    right:
        arm_id: panda_1
        joint_names:
            - panda_1_joint1
            - panda_1_joint2
            - panda_1_joint3
            - panda_1_joint4
            - panda_1_joint5
            - panda_1_joint6
            - panda_1_joint7
    left:
        arm_id: panda_2
        joint_names:
            - panda_2_joint1
            - panda_2_joint2
            - panda_2_joint3
            - panda_2_joint4
            - panda_2_joint5
            - panda_2_joint6
            - panda_2_joint7

teleop_joint_pd_example_controller:
   type: franka_example_controllers/TeleopJointPDExampleController
   leader:
       arm_id: panda_1
       joint_names:
            - panda_1_joint1
            - panda_1_joint2
            - panda_1_joint3
            - panda_1_joint4
            - panda_1_joint5
            - panda_1_joint6
            - panda_1_joint7
       # WARNING: Varying the following parameters may result in instability!
       d_gains: [1.0, 1.0, 1.0, 1.0, 0.33, 0.33, 0.17]
       contact_force_threshold: 4.0 # [N]
   follower:
       arm_id: panda_2
       joint_names:
            - panda_2_joint1
            - panda_2_joint2
            - panda_2_joint3
            - panda_2_joint4
            - panda_2_joint5
            - panda_2_joint6
            - panda_2_joint7
       # WARNING: Varying the following parameters may result in instability!
       p_gains: [900.0, 900.0, 900.0, 900.0, 375.0, 225.0, 100.0]
       d_gains: [45.0, 45.0, 45.0, 45.0, 15.0, 15.0, 10.0]
       drift_comp_gains: [4.3, 4.3, 4.3, 4.3, 4.3, 4.3, 4.3]
       dq_max_lower: [0.8, 0.8, 0.8, 0.8, 2.5, 2.5, 2.5] # [rad/s]
       dq_max_upper: [2.0, 2.0, 2.0, 2.0, 2.5, 2.5, 2.5] # [rad/s]
       ddq_max_lower: [0.8, 0.8, 0.8, 0.8, 10.0, 10.0, 10.0] # [rad/s²]
       ddq_max_upper: [6.0, 6.0, 6.0, 6.0, 15.0, 20.0, 20.0] # [rad/s²]
       contact_force_threshold: 5.0 # [N]

position_force_hybird_controller:
    type: franka_example_controllers/position_force_hybird_controller
    arm_id: panda
    joint_names:
        - panda_joint1
        - panda_joint2
        - panda_joint3
        - panda_joint4
        - panda_joint5
        - panda_joint6
        - panda_joint7

cartesian_pose_topic_controller:
    type: franka_example_controllers/CartesianPoseControllerTopic
    arm_id: panda

cartesian_pose_ZJK_controller:
    type: franka_example_controllers/CartesianPoseZJKController
    arm_id: panda

cartesian_position_force_hybird_controller:
    type: franka_example_controllers/cartesian_position_force_hybird_controller
    arm_id: panda
    joint_names:
        - panda_joint1
        - panda_joint2
        - panda_joint3
        - panda_joint4
        - panda_joint5
        - panda_joint6
        - panda_joint7

joint_impedance_example_controller_topic:
    type: franka_example_controllers/JointImpedanceExampleControllerTopic
    arm_id: panda
    joint_names:
        - panda_joint1
        - panda_joint2
        - panda_joint3
        - panda_joint4
        - panda_joint5
        - panda_joint6
        - panda_joint7
    k_gains:
        - 600.0
        - 600.0
        - 600.0
        - 600.0
        - 250.0
        - 150.0
        - 50.0
    d_gains:
        - 50.0
        - 50.0
        - 50.0
        - 20.0
        - 20.0
        - 20.0
        - 10.0
    radius: 0.1
    acceleration_time: 2.0
    vel_max: 0.15
    publish_rate: 10.0
    coriolis_factor: 1.0

joint_position_example_topic_controller:
    type: franka_example_controllers/JointPositionExampleTopicController
    joint_names:
        - panda_joint1
        - panda_joint2
        - panda_joint3
        - panda_joint4
        - panda_joint5
        - panda_joint6
        - panda_joint7

cartesian_velocity_Z_impedance_LiMD:
  type: franka_example_controllers/CartesianVelocityZImpedanceLiMD
  arm_id: panda


panda_leader_cartesian_impedance_controller:
  type: franka_example_controllers/CartesianImpedanceExampleController
  # type: thorvald_panda_controllers/LeaderCartesianImpedanceController
  arm_id: panda
  joint_names:
      - panda_joint1
      - panda_joint2
      - panda_joint3
      - panda_joint4 
      - panda_joint5
      - panda_joint6
      - panda_joint7
